StandardLibraryEssentials/StandardTemplateLibrary/UnorderedContainers/task.md

```diff
@@ -0,0 +1,96 @@
+Is there a way to create a container that has an `O(1)` complexity for insertion, deletion, and search? Sounds like magic, right?
+Well, it is not. STL provides `std::unordered_set` and `std::unordered_map` containers that are implemented using [hash tables](https://en.wikipedia.org/wiki/Hash_table).
+
+What distinguishes regular `std::set` and `std::map` from their unordered counterparts? Due to the implementation details (which will be explained further), their main difference is the lack of a stable order of elements. There are iterators for such containers, but as you traverse the container using these iterators, you will most likely notice that the elements are not be sorted in ascending order.
+This, however, is a small price to pay for the `O(1)` complexity of the search, insertion, and deletion operations. 
+
+The main methods of `std::unordered_set` and `std::unordered_map` include:
+* `insert(value)` - inserts the element `value` into the container
+* `emplace(args...)` - constructs the element in-place
+* `erase(value)` - removes the element `value` from the container
+* `find(value)` - returns an iterator to the element `value` if it is found, otherwise, it returns `end()`
+
+To be fair, all these methods have **amortized** `O(1)` complexity, since sometimes hashes of different keys can be the same. This is called a [collision](https://en.wikipedia.org/wiki/Hash_table#Collision_resolution). In such situations, the whole table needs to be rehashed, which is the reason why the methods of an unordered container can perform in `O(n)` time in the worst-case scenario. 
```

Comment:
Move this paragraph after the next two paragraphs (about hash table and `std::hash`), and put it under a hint.